### PR TITLE
Twitter clone

### PR DESCRIPTION
**Pull request by tmampa**
- Created a clone of Twitter using Ruby on Rails showcasing my knowledge on authentication.
- Created using ruby on rails
- authentication using Devise gem
- Page styling using the Bulma framework
- 
##  Fixed changes
- User can now see authors only when signed in
- Readme updated
- Routes changed

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -1,3 +1,6 @@
 Rails.application.routes.draw do
+  devise_for :users, controllers: { registrations: 'registrations' }
+  resources :tweets
```
Comment (rindrajosia) [inline config/routes.rb:3]: It creates the application routes using resources :tweets, only: [:new, :create, :index]

File: app/controllers/tweets_controller.rb [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    @tweets = Tweet.all.order('created_at DESC')
+    @tweet = Tweet.new
+  end
+
+  # GET /tweets/1 or /tweets/1.json
+  def show; end
+
+  # GET /tweets/new
+  def new
+    @tweet = current_user.tweets.build
```
Comment (rindrajosia) [inline app/controllers/tweets_controller.rb]: it initializes an empty post @post = Post.new

File: app/views/layouts/application.html.erb [hunk 3]
```diff
@@ -1,16 +1,65 @@
...
+                <%= link_to 'Sign Out', destroy_user_session_path, method: :delete, class: 'button is-info' %>
+              </p>
+            <% else %>
+              <p class="control">
+                <%= link_to 'Sign In', new_user_session_path, class: 'button is-info' %>
+              </p>
+              <p class="control">
+                <%= link_to 'Sign Up', new_user_registration_path, class: 'button is-info' %>
+              </p>
+            <% end %>
```
Comment (rindrajosia) [inline app/views/layouts/application.html.erb:46]: Please remove logic in your view

File: app/views/layouts/application.html.erb [hunk 4]
```diff
@@ -1,16 +1,65 @@
...
+    <% if flash[:notice] %>
+      <div class="notification is-primary global-notification">
+        <p class="notice"><%= notice %></p>
+      </div>
+    <% end %>
+    <% if flash[:alert] %>
+      <div class="notification is-danger global-notification">
+        <p class="alert"><%= alert %></p>
+      </div>
+    <% end %>
```
Comment (rindrajosia) [inline app/views/layouts/application.html.erb:61]: Please remove logic in your view

File: app/views/tweets/_feed.html.erb [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    <figure class="media-left">
+      <p class="image is-64x64">
+        <%= gravatar_image_tag(current_user.email, size: 64, alt: current_user.name) %>
+      </p>
+    </figure>
+    <div class="media-content">
+      <%= render 'tweets/form' %>
+    </div>
+  </article>
+<% end %>
```
Comment (rindrajosia) [inline app/views/tweets/_feed.html.erb:13]: Please remove logic in your view

File: app/views/tweets/_feed.html.erb [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+              <%= link_to tweet, method: :delete, data: { confirm: 'Are you sure you want to delete this tweet?' }, class: 'level-item' do %>
+                <span class="icon"><i class="far fa-trash-alt" aria-hidden="true"></i></span>
+              <% end %>
+            </div>
+          </nav>
+          <% end %>
+        </div>
+      </article>
+    </div>
+  <% end %>
```
Comment (rindrajosia) [inline app/views/tweets/_feed.html.erb]: Please remove logic in your view

File: app/views/tweets/index.html.erb [hunk 7]
```diff
@@ -0,0 +1,13 @@
+<section class="section">
+  <div class="container">
+    <div class="columns">
+      <% if user_signed_in? %>
+        <%= render 'profile' %>
+      <% else %>
+        <%= render 'trends' %>
+      <% end %>
+      <%= render 'feed' %>
+      <%= render 'who-to-follow' %>
```
Comment (rindrajosia) [inline app/views/tweets/index.html.erb:10]: Please remove logic in your view

**Review by rindrajosia — CHANGES_REQUESTED** [review]
Changes Required :red_circle:
To Highlight: :angel:

- It uses the devise gem for authentication :+1: 
- User can sign in :100: 
- Signed-in user can create new post :+1: 
- Signed-in user can see post's author :100: 
- User can sign_out :100: 

To improve :factory_worker: [REQUIRED]:

- [x] Please use tools like Grammarly to help you with your readme as it is the first thing users check.
- [x] Guest should see only post's text ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/guest_can_see_only_posts_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/authentication))
- [x] It creates the application routes using `resources :posts, only: [:new, :create, :index]` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/posts_routes_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/authentication))
- [x] **PostsController # new**: it initializes an empty post `@post = Post.new` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/posts_new_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/authentication))

Nice to have [OPTIONAL] :

- [ ] Please remove logic in your view. You can use helpers for it.

That's it, you can submit another code review request after making these changes :recycle: :innocent:
With all my respect :pray:
Rindra Josia

File: app/controllers/tweets_controller.rb (hunk 2 above)
Comment (afanimartin) [reply app/controllers/tweets_controller.rb]: - Initialize an empty tweet: @tweet = Tweet.new

**Review by afanimartin — CHANGES_REQUESTED** [review]
## Changes Required 🙏🏿 🙏🏿 🙏🏿 
Nice work making the requested changes. A guest user can only see post text. Well done.
You, however, missed fixing a change requested earlier on. Check my comment for suggestions.


### Martin Afani
#### Slack: whiz25

**Review by george-shammar — APPROVED** [review]
### **STATUS: APPROVED** :heavy_check_mark: :heavy_check_mark: 

Thank you for making the required change.

Your repo looks professional.

I'm happy to approve your project.

I wish you good luck  with your future projects.

Cheers :smile:

**Pull request merged**